felix/nftables/maps.go

```diff
@@ -0,0 +1,677 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package nftables
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sirupsen/logrus"
+	"sigs.k8s.io/knftables"
+
+	"github.com/projectcalico/calico/felix/deltatracker"
+	"github.com/projectcalico/calico/felix/ipsets"
+	"github.com/projectcalico/calico/felix/logutils"
+	"github.com/projectcalico/calico/libcalico-go/lib/set"
+)
+
+var gaugeVecNumMaps = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+	Name: "felix_nft_maps",
+	Help: "Number of active Calico nftables maps.",
+}, []string{"ip_version"})
+
+type MapType string
+
+const MapTypeInterfaceMatch MapType = "interfaceMatch"
+
+type MapsDataplane interface {
+	AddOrReplaceMap(meta MapMetadata, members map[string][]string)
+	RemoveMap(setID string)
+
+	MapUpdates() *mapUpdates
+	FinishMapUpdates(updates *mapUpdates)
+	LoadDataplaneState() error
+}
+
+var _ MapsDataplane = &Maps{}
+
+type MapMetadata struct {
+	ID   string
+	Type MapType
+}
+
+// Maps manages a whole "plane" of maps, i.e. all the IPv4 maps, or all the IPv6 maps.
+type Maps struct {
+	IPVersionConfig *ipsets.IPVersionConfig
+
+	// mapNameToAllMetadata contains an entry for each map that has been
+	// added by a call to AddOrReplaceMap (and not subsequently removed).
+	mapNameToAllMetadata map[string]MapMetadata
+
+	// mapNameToProgrammedMetadata tracks the maps that we want to program and
+	// those that are actually in the dataplane.  It's Desired() map is the
+	// subset of mapNameToAllMetadata.
+	// Its Dataplane() map contains all maps matching the IPVersionConfig
+	// that we think are in the dataplane.  This includes any temporary IP
+	// maps and maps that we discovered on a resync (neither of which will
+	// have entries in the Desired() map).
+	mapNameToProgrammedMetadata *deltatracker.DeltaTracker[string, MapMetadata]
+
+	// mainSetNameToMembers contains entries for all maps that are in
+	// mapNameToAllMetadata along with entries for "main" (non-temporary)
+	// maps that we think are still in the dataplane. For maps that are in mapNameToAllMetadata, the
+	// Desired() side of the tracker contains the members that we've been told
+	// about.  Otherwise, Desired() is empty.  The Dataplane() side of the
+	// tracker contains the members that are thought to be in the dataplane.
+	mainSetNameToMembers map[string]*deltatracker.SetDeltaTracker[MapMember]
```

Comment:
I think using a `SetDeltaTracker` is subtly broken here.  If the value of the KV pair changes then that will be a distinct entry in the set, but only one entry in the kernel due to the shared key.  That matters when loading data from the kernel, you'll get two entries in the Dataplane side of the tracker instead of one and you'll queue up an add + deletion instead of an update.

Only spotted it because I started looking into https://github.com/projectcalico/calico/issues/9450 today and I fell into a very similar trap there.

Reply:
@fasaxc I'm not sure that's actually a problem in this code though? Maybe I am misunderstanding, but sounds like it will generate both an add + delete, but since we program those in the same nftables transaction, and also order them such that the delete comes before the add, the dataplane should be updated correctly.

Reply:
Ah yep, ordering the updates covers the dataplane side of it, I think.  The VXLAN FDB did the updates the other way around.

Could we trip ourselves up on the Desired side of the map in future?  I'm thinking that the caller changes the verdict that they want and that results in both `key -> verdict1` and `key -> verdict2` in the set (when the dataplane can only handle one of those at a time).

Reply:
Answer is no right now because the API doesn't allow individual updates, I think.